### PR TITLE
feature/138

### LINKED ISSUE #138: Allow optional id prefixes in BaseModel and BaseCollection
_From @n8rzz on January 7, 2017 5:11_

Each class is an inheritable and uses `_uniqueId()` to generate a value for `_id`. This task would allow for an optional prefix to that property.

- [x] support a string prefix with output like `positionModel-42`
- [x] support no-prefix where default output is `baseModel-42`

_Copied from original issue: n8rzz/atc#255_

### PR DESCRIPTION
**Pull request by Dabea**
Resolves #138.

Add additional context to the  ._id property by accepting an optional parameter as a string
Also if given an object will be able to pull some information as well like airport name or fix.

### CODE REVIEW

File: src/assets/scripts/client/base/BaseModel.js [hunk 1]
```diff
@@ -45,4 +45,27 @@ export default class BaseModel {
...
     }
+
+    /**
+     * This will verify if the given argument is a string and will return a modified pre-value for the _id
+     *
+     * @for BaseModel
+     * @method veriftyModelName
+     * @param {string / Object}
+     */
+    verifyModelName(modelName) {
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: are we _verifying_ a model name here?  or are we _building_ something?

File: test/base/BaseModel.spec.js [hunk 2]
```diff
@@ -4,10 +4,49 @@ import _isString from 'lodash/isString';
...
+});
+
+ava('instantiates with an _id with string property', t => {
+    const resultString = new BaseModel('string');
+
+    t.true(_isString(resultString._id));
+});
+
+ava('instantiates with an _id with an Object', t => {
+    const resultObj = new BaseModel({});
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: this should only ever accept a string, this test can be removed

File: test/base/BaseModel.spec.js [hunk 3]
```diff
@@ -4,10 +4,49 @@ import _isString from 'lodash/isString';
...
+
+    t.true(_isString(resultObj._id));
+});
+
+ava('instantiates with an _id with an array', t => {
+    const resultArray = new BaseModel([]);
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: this should only ever accept a string, this test can be removed

File: test/base/BaseModel.spec.js [hunk 4]
```diff
@@ -4,10 +4,49 @@ import _isString from 'lodash/isString';
...
+
+    t.true(_isString(resultArray._id));
+});
+
+ava('instantiates with an _id with an boolean', t => {
+    const resultBol = new BaseModel(false);
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: this should only ever accept a string, this test can be removed

File: test/base/BaseModel.spec.js [hunk 5]
```diff
@@ -4,10 +4,49 @@ import _isString from 'lodash/isString';
...
+
+    t.true(_isString(resultBol._id));
+});
+
+ava('instantiates with an _id with an int', t => {
+    const resultInt = new BaseModel(42);
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: this should only ever accept a string, this test can be removed

File: src/assets/scripts/client/base/BaseModel.js [hunk 6]
```diff
@@ -45,4 +45,27 @@ export default class BaseModel {
...
+        if (typeof modelName === 'string') {
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: there is a lodash helper for this `_isString()`

File: src/assets/scripts/client/base/BaseModel.js [hunk 7]
```diff
@@ -8,15 +8,15 @@ import _uniqueId from 'lodash/uniqueId';
...
+    constructor(modelName = 'BaseModel') {
         /**
          * Unigue string id that can be used to differentiate this model instance from another.
          *
          * @property _id
          * @type {string}
          * @private
          */
-        this._id = _uniqueId();
+        this._id = this.verifyModelName(modelName) + _uniqueId();
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: additionally, `_uniqueId` accepts a string that can be used as the id base.  so in our case: 
```
_uniqueId(`${modelName}-`)
```

File: test/base/BaseModel.spec.js (hunk 3 above)
Comment (Dabea) [inline test/base/BaseModel.spec.js]: There are cases in the code where it is currently being called by other parameters. I was confused on why that was the case though.

File: src/assets/scripts/client/base/BaseModel.js [hunk 8]
```diff
@@ -45,4 +45,27 @@ export default class BaseModel {
...
+     * @method veriftyModelName
+     * @param {string / Object}
+     */
+    verifyModelName(modelName) {
+        if (typeof modelName === 'string') {
+            return `${modelName}-`;
+        }
+
+        // This will give more infomation on airpot names and runways to help identify if there is an issue.
+        if (modelName.name === undefined && typeof modelName === 'object') {
```
Comment (Dabea) [inline src/assets/scripts/client/base/BaseModel.js]: This will show additional information on the object that is used to create it give you `_id`  values like 
`BaseModel-Caribbean Airlines-1087`  So you can see if it is only happening with certain Data, But I can just remove it.
I did an oversimplification when i was cleaning up the logic here
This would be a more accurate version
```
if (modelName.name === undefined) {
            modelName.name = '';
        }

        // This will give more information on airport names and runways to help identify if there is an issue.
        if (typeof modelName === 'object') {
            console.log(modelName);
            return `BaseModel-${modelName.name}-`;
        }
```

File: test/base/BaseModel.spec.js (hunk 3 above)
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: hmm, the inheriting classes should not need to send their parameters to the parent with `super(value, value, value)` etc.

If you find any of those, go ahead and remove them (carefully).  Any class inheriting `BaseModel` should have an empty call to `super()`.

File: src/assets/scripts/client/base/BaseModel.js (hunk 8 above)
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: I see what you're doing there.  That actually isn't the intent.  I may not have explained the terms here very well (and they do collide a little, which is confusing).

This class, and all classes inheriting it, has an `_id` property.  Without any input from us, this `_id` will be a string of numbers that increments every time this class is instantiated.  so you'll end up with `_id`s that look like: `this._id = 3432` and so on.  

My intent for this is for an inheriting class to have the ability to send an optional sting that can be added to the `_id` that helps differentiate each instance.  so say we have the `FixModel`, which inherits the `BaseModel`.  In the `FixModel` constructor, we would call super with that optional `_id` string : `super('FixModel')` which would then be added to that id.  this would give us an `_id` of `FixModel-3432`.

I think the _model_ you are thinking of here is actually the `AircraftTypeDefinitionModel` or the `AirlineModel`, those are different things entirely.  So this block can go away.  It was a good thought, but thats not what I'm looking for here.

File: src/assets/scripts/client/base/BaseModel.js [hunk 9]
```diff
@@ -47,23 +49,17 @@ export default class BaseModel {
...
      *
      * @for BaseModel
      * @method veriftyModelName
      * @param {string / Object}
      */
-    verifyModelName(modelName) {
-        if (typeof modelName === 'string') {
+    buildModelPrefix(modelName) {
+        if (_isString(modelName)) {
             return `${modelName}-`;
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: This if logic should be reversed.  We should be exclusive, not inclusive, with our logic blocks.
```javascript
// instead of (inclusive)
if (_isString(modelName)) { 
  return `${modelName}-`;
}

// we should do (exclusive)
if (!_isString(modelName)) {
  return `${modelName}-`;
}
``` 
a method should eject (return early) as soon as we find data we don't want.

Also, since `modelName` is defaulted to `BaseModel` in the constructor, we shouldn't need to return that string again below.  We can just use `modelName-`.

So with that in mind, does the value we return change?  Doesn't look like it.  But we do need to do something in the case we receive bad data, and handle that appropriately.  In this case, we should use `console.error` if we receive a param that is not a string, but still return our `modelName-` because that kind of error shouldn't prevent the model from instantiating.

File: src/assets/scripts/client/base/BaseModel.js [hunk 10]
```diff
@@ -8,15 +9,16 @@ import _uniqueId from 'lodash/uniqueId';
...
+    constructor(modelName = 'BaseModel') {
         /**
          * Unigue string id that can be used to differentiate this model instance from another.
          *
          * @property _id
          * @type {string}
          * @private
          */
-        this._id = _uniqueId();
+        const optionalModelPrefix = this.buildModelPrefix(modelName);
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: this line should be moved to just under the constructor declaration

File: CHANGELOG.md [hunk 11]
```diff
@@ -13,8 +13,8 @@
 
 
 ### Features
-
-
+- Fixes an instance of two runways called "34R" in Shanghai Pudong [#149](https://github.com/openscope/openscope/issues/149)
```
Comment (Dabea) [inline CHANGELOG.md]: tried git pull origin develop then pushed the changes however this diff did not go away

File: src/assets/scripts/client/base/BaseModel.js [hunk 12]
```diff
@@ -8,15 +8,15 @@ import _uniqueId from 'lodash/uniqueId';
  * @class BaseModel
  */
 export default class BaseModel {
-    constructor() {
+    constructor(modelName = 'BaseModel') {
```
Comment (Dabea) [inline src/assets/scripts/client/base/BaseModel.js]: Added tests to check the default param

File: src/assets/scripts/client/base/BaseModel.js [hunk 13]
```diff
@@ -45,4 +49,22 @@ export default class BaseModel {
...
+     * This will verify if the given argument is a string otherwise it will return 'Base Model'
+     *
+     * @for BaseModel
+     * @method veriftyModelName
+     * @param {string / Object}
+     */
+    buildIdPrefix(modelName) {
+        if (!_isString(modelName)) {
+            console.error('BaseModel#constructor expects a string for its first paramert but a string was not given');
+            return 'BaseModel-';
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: the `console.error` should be a `throw` instead.

since `modelName` is defaulted to `BaseModel`, the return here is redundant and can be removed.

File: src/assets/scripts/client/base/BaseModel.js [hunk 14]
```diff
@@ -45,4 +49,22 @@ export default class BaseModel {
...
     }
+
+    /**
+     * This will verify if the given argument is a string otherwise it will return 'Base Model'
+     *
+     * @for BaseModel
+     * @method veriftyModelName
+     * @param {string / Object}
+     */
+    buildIdPrefix(modelName) {
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: could we make this method private?

File: src/assets/scripts/client/base/BaseModel.js (hunk 13 above)
Comment (Dabea) [inline src/assets/scripts/client/base/BaseModel.js]: So I should change it to 
`return `${modelName}-` 

Or should the default be changed to 'BaseModel-'

Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: `return ${modelName}`;

Comment (Dabea) [inline src/assets/scripts/client/base/BaseModel.js]: Just to be clear is the desired out put to look like
BaseModel-1234   Or
BaseMode1234

Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: correctL `BaseModel-1234` will be the end result.  and since `_uniqueId()` takes care of the `1234` part of it, we only need the `BaseModel-` part.

File: src/assets/scripts/client/base/BaseModel.js (hunk 14 above)
Comment (Dabea) [inline src/assets/scripts/client/base/BaseModel.js]: By private do you mean just giving it a naming scheme with an _ in front of the method name and adding @private in the doc blocks?

File: src/assets/scripts/client/base/BaseModel.js (hunk 13 above)
Comment (Dabea) [inline src/assets/scripts/client/base/BaseModel.js]: So then the default parameter should be  changed to `BaseModel-`? and just return ${modelName}?

File: src/assets/scripts/client/base/BaseModel.js (hunk 14 above)
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: that is correct 👍

File: src/assets/scripts/client/base/BaseModel.js (hunk 13 above)
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: no, the default should stay the same.  this is the correct way to do it, by adding the `-` in the return.

Comment (Dabea) [inline src/assets/scripts/client/base/BaseModel.js]: Also if I do a `throw` then it will never hit the return statement and things will break? (not sure on the term obtrusive error handling?) Or should this be non-obtrusive and just give a default response?

Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: Yep, we want them to break here because the developer is doing something they shouldn't be doing. 

See my first comment about the return, the throw should be the only thing in this block.

**Comment by Dabea** [comment]
Earlier you had me change the error to obtrusive error handling however we previously built theses tests with Ava. Should I remove theses tests or just change them to fail on everything  but pass when a string or nothing is given?

```ava('sets the #_id when passed invalid parameters', (t) => {
    const modelojb = new BaseModel({});
    t.true(modelojb._id.indexOf('BaseModel') !== -1);

    const modelArray = new BaseModel([]);
    t.true(modelArray._id.indexOf('BaseModel') !== -1);

    const modelInt = new BaseModel(42);
    t.true(modelInt._id.indexOf('BaseModel') !== -1);

    const modelBool = new BaseModel(false);
    t.true(modelBool._id.indexOf('BaseModel') !== -1);
});

ava('makes sure the model in instantiates with any data type', t => {
    t.notThrows(() => new BaseModel());
    t.notThrows(() => new BaseModel([]));
    t.notThrows(() => new BaseModel({}));
    t.notThrows(() => new BaseModel(42));
    t.notThrows(() => new BaseModel('string'));
    t.notThrows(() => new BaseModel(false));
});
```

File: src/assets/scripts/client/navigationLibrary/StandardRoute/StandardRouteModel.js [hunk 15]
```diff
@@ -156,7 +157,7 @@ export default class StandardRouteModel extends BaseModel {
          */
         this._entryCollection = null;
 
-        return this._init(standardRoute);
```
Comment (n8rzz) [inline src/assets/scripts/client/navigationLibrary/StandardRoute/StandardRouteModel.js]: why did you remove this return?

File: test/base/BaseModel.spec.js [hunk 16]
```diff
@@ -4,10 +4,39 @@ import _isString from 'lodash/isString';
 import BaseModel from '../../src/assets/scripts/client/base/BaseModel';
 import ExtendedBaseModelFixture from './_fixtures/ExtendedBaseModelFixture';
 
+ava('sets the #_id when passed valid parameters', (t) => {
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: this test can be improved a bit.
1) if you're going to do multiple assertions inside of a test you should use a single var to hold the results:
```javascript
let model = new BaseModel();
// ...
model = new BaseModel('string');
```

2) use a more realistic example for the passed model name:
```javascript
// instead of
modelBool = new BaseModel('string');

// do
modelBool = new BaseModel('modelName');
```  This adds a little more context to the test and informs a future developer about the intent.

File: test/base/BaseModel.spec.js [hunk 17]
```diff
@@ -4,10 +4,39 @@ import _isString from 'lodash/isString';
...
+    const modelInt = new BaseModel();
+    t.true(modelInt._id.indexOf('BaseModel') !== -1);
+
+    const modelBool = new BaseModel('string');
+    t.true(modelBool._id.indexOf('string') !== -1);
+    console.log(modelBool._id);
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: errant console.log

File: test/base/BaseModel.spec.js [hunk 18]
```diff
@@ -4,10 +4,39 @@ import _isString from 'lodash/isString';
...
+});
+
+ava('makes sure the model fails to instantiate with any incorrect', t => {
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: take a look at some of the other test files and try to stay close to the convention established there with how this type of test is worded.  its usually along the lines of: `throws when passed invalid parameters` or something like that.

File: test/base/BaseModel.spec.js [hunk 19]
```diff
@@ -4,10 +4,39 @@ import _isString from 'lodash/isString';
...
+    t.notThrows(() => new BaseModel());
+    t.throws(() => new BaseModel([]));
+    t.throws(() => new BaseModel({}));
+    t.throws(() => new BaseModel(42));
+    t.notThrows(() => new BaseModel('string'));
+    t.throws(() => new BaseModel(false));
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: the next test validates the `notThrows` cases, so they can be removed from this test.

File: test/base/BaseModel.spec.js [hunk 20]
```diff
@@ -4,10 +4,39 @@ import _isString from 'lodash/isString';
...
+});
+
+ava('makes sure the model in instantiates with the correct data type', t => {
+    t.notThrows(() => new BaseModel());
+    t.notThrows(() => new BaseModel('string'));
+});
+
 ava('instantiates with an _id property', t => {
-    const result = new BaseModel();
+    const resultEmpty = new BaseModel();
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: be sure to use `model` as the const name.  this is just an [undocumented] convention we've established in our tests to make it more readable.

File: test/base/BaseModel.spec.js [hunk 21]
```diff
@@ -4,10 +4,39 @@ import _isString from 'lodash/isString';
...
+
+    t.true(_isString(resultEmpty._id));
+});
+
+ava('instantiates with an _id with string property', t => {
```
Comment (n8rzz) [inline test/base/BaseModel.spec.js]: this test is redundant and can be removed

File: src/assets/scripts/client/navigationLibrary/StandardRoute/StandardRouteModel.js (hunk 15 above)
Comment (Dabea) [inline src/assets/scripts/client/navigationLibrary/StandardRoute/StandardRouteModel.js]: I was tracking down a bug,  Accidentally left it out after it was found. Adding it back in. I Should have commented it out and added the `this._init(standardRoute);` as its own line with a comment showing that though

File: test/base/BaseModel.spec.js (hunk 16 above)
Comment (Dabea) [inline test/base/BaseModel.spec.js]: Made the changes now uses a common variable name and the given string is now modelName

File: src/assets/scripts/client/base/BaseModel.js [hunk 22]
```diff
@@ -45,4 +49,22 @@ export default class BaseModel {
...
+     * This will verify if the given argument is a string otherwise it will return 'Base Model'
+     *
+     * @for BaseModel
+     * @method veriftyModelName
+     * @param {string / Object}
+     * @private
+     */
+    _buildIdPrefix(modelName) {
+        if (!_isString(modelName)) {
+            throw new TypeError('BaseModel#constructor expects a string for its first paramert but a string was not given');
```
Comment (n8rzz) [inline src/assets/scripts/client/base/BaseModel.js]: paramert -> parameter

File: CHANGELOG.md [hunk 23]
```diff
@@ -68,8 +68,11 @@
 
 ### Features
 - Enumerate magic number in RunwayModel [#269](https://github.com/openscope/openscope/issues/269)
+
+- Adds more contex to the Model classes by adding an optional input paramater [#138](https://github.com/openscope/openscope/issues/138)
```
Comment (n8rzz) [inline CHANGELOG.md]: make sure this is in the correct spot in the current chagnelog.  this diff looks like its from pre v5.0.0

Comment (erikquinn) [inline CHANGELOG.md]: Fixed.
